l3kernel/l3tl.dtx

```diff
@@ -2271,26 +2327,26 @@
 % \begin{macro}[EXP,pTF]{\tl_if_single:n}
 % \begin{macro}[EXP,pTF]{\@@_if_single:n}
 %   This test is similar to \cs{tl_if_empty:nTF}.  Expanding
-%   \cs{use_none:nn} |#1| |??| once yields an empty result if |#1| is
+%   \cs{use_none:n} |#1| |?| once yields an empty result if |#1| is
 %   blank, a single~|?| if |#1| has a single item, and otherwise yields
-%   some tokens ending with |??|.  Then, \cs{tl_to_str:n} makes sure
+%   some tokens ending with |?|.  Then, \cs{__kernel_tl_to_str:w} makes sure
 %   there are no odd category codes.  An earlier version would compare
 %   the result to a single~|?| using string comparison, but the Lua call
 %   is slow in \LuaTeX{}.  Instead, \cs{@@_if_single:nnw} picks the
 %   second token in front of it.  If |#1| is empty, this token is
-%   the trailing~|?| and the catcode test yields \texttt{false}.  If
-%   |#1| has a single item, the token is~|^| and the catcode test
+%   the trailing~|?| and the |\if:w| test yields \texttt{false}.  If
+%   |#1| has a single item, the token is~|\scan_stop:| and the |\if:w| test
 %   yields \texttt{true}.  Otherwise, it is one of the characters
-%   resulting from \cs{tl_to_str:n}, and the catcode test yields
-%   \texttt{false}.  Note that \cs{if_catcode:w} and
+%   resulting from \cs{tl_to_str:n}, and the |\if:w| test yields
+%   \texttt{false}.  Note that \cs{if:w} and
 %   \cs{__kernel_tl_to_str:w} are primitives that take care of
 %   expansion.
 %    \begin{macrocode}
 \prg_new_conditional:Npnn \tl_if_single:n #1 { p , T , F , TF }
   {
-    \if_catcode:w ^ \exp_after:wN \@@_if_single:nnw
+    \if:w \scan_stop: \exp_after:wN \@@_if_single:nnw
         \__kernel_tl_to_str:w
-          \exp_after:wN { \use_none:nn #1 ?? } ^ ? \s_@@_stop
+          \exp_after:wN { \use_none:n #1 ? } \scan_stop: ? \s_@@_stop
```

Comment:
This optimization is wrong because a token list that is not blank may still have a string representation made of spaces: this happens for the control space <code>\ </code> when the escapechar is not printable; it also happens for space characters with weird category codes.  With your code, if what follows the first item in `#1` has a blank string representation, then it will be ignored by `\@@_if_single:nnw` just as if it were really blank in the token list.  The code will then deem the token list blank.  An example, which gives `\ERROR` with your new code:

    \char_set_lccode:nn { `a } { 32 }
    \lowercase { \tl_if_single:nT { b a } } { \ERROR }

On the other hand, I think `\@@_if_single:nnw` can be replaced by `\use_none:n` and `\scan_stop: ? \s_@@_stop` by `\prg_do_nothing: \scan_stop:`, but I am not sure whether this speeds things up (it replaces a three-parameter macro by a one-parameter one `\use_none:n` and a zero-parameter one `\prg_do_nothing:` in the true case).

Reply:
Replacing `\@@_if_single:nnw` with `\use_none:n` and putting `\prg_do_nothing: \scan_stop:` there would result in the test being true for a blank argument, so doesn't work as well.

Reply:
Oops, correct, so I guess the original code may be optimal.